uCNC/src/hal/mcus/samd21/mcu_samd21.c

```diff
@@ -1326,6 +1483,105 @@ uint8_t mcu_spi_xmit(uint8_t c)
 	return (uint8_t)SPICOM->SPI.DATA.reg;
 }
 
+void SPI_ISR()
+{
+	if(SPICOM->SPI.INTFLAG.bit.DRE && spi_tx_length > 0)
+	{
+		// Send next byte
+		SPICOM->SPI.DATA.reg = *spi_tx_buffer++;
+		--spi_tx_length;
+	}
+	if(SPICOM->SPI.INTFLAG.bit.RXC && spi_rx_length > 0)
+	{
+		// Store received byte
+		*spi_rx_buffer++ = SPICOM->SPI.DATA.reg;
+		--spi_rx_length;
+	}
+	if(spi_tx_length == 0 && spi_rx_length == 0)
+	{
+		// Transfer complete
+		spi_port_state = SPI_TRANSMIT_FINISHED;
+		// Disable interrupts
+		SPICOM->SPI.INTENCLR.bit.DRE = 1;
+		SPICOM->SPI.INTENCLR.bit.RXC = 1;
+	}
+	NVIC_ClearPendingIRQ(SPI_IRQ);
+}
+
+bool mcu_spi_bulk_transfer(const uint8_t *tx_data, uint8_t *rx_data, uint16_t datalen)
+{
+	if(!spi_enable_dma)
+	{
+		// Bulk transfer without DMA
+		if(spi_port_state == SPI_IDLE)
+		{
+			spi_tx_buffer = tx_data;
+			spi_tx_length = datalen;
+			if(rx_data)
+			{
+				spi_rx_buffer = rx_data;
+				spi_rx_length = datalen;
+			}
+			else
+			{
+				spi_rx_data = 0;
```

Comment:
There is a typo that results in a compilation error. Should be `spi_rx_buffer` not `spi_rx_data`